firedrake/slate/slac/compiler.py

```diff
@@ -530,6 +590,325 @@ def parenthesize(arg, prec=None, parent=None):
+        idx = builder.gem_indices[str(builder.expression)+"out"]
+    else:
+        idx = traversed_gem_expr_dag[0].multiindex
+    ret_vars = [gem.Indexed(gem.Variable("output", shape), idx)]
+
+    # TODO the global argument generation must be made nicer
+    # Maybe this can be done in the builder?
+    if len(builder.args_extents) > 0:
+        arg = loopy.GlobalArg("coords", shape=(builder.args_extents[builder.coordinates_arg],), dtype="double")
+        args.append(arg)
+    else:
+        arg = loopy.GlobalArg("coords", shape=(builder.expression.shape[0],), dtype="double")
+        args.append(arg)
+
+    if builder.needs_cell_orientations:
+        args.append(loopy.GlobalArg("orientations", shape=(builder.args_extents[builder.cell_orientations_arg],), dtype=np.int32))
+
+    if builder.needs_cell_sizes:
+        args.append(loopy.GlobalArg("cell_sizes", shape=(builder.args_extents[builder.cell_size_arg],), dtype="double"))
+
+    # Add coefficients, where AssembledVectors sit on top to args.
+    # The fact that the coefficients need to go into the order of
+    # builder.expression.coefficients(), plus mixed coefficients,
+    # plus split always generating new functions, makes life a bit harder.
+    coeff_shape_list = []
+    coeff_function_list = []
+    for v in builder.coefficient_vecs.values():
+        for coeff_info in v:
+            coeff_shape_list.append(coeff_info.shape)
+            coeff_function_list.append(coeff_info.vector._function)
+    get = 0
+    for c in builder.expression.coefficients():
+        try:
+            indices = [i for i, x in enumerate(coeff_function_list) if x == c]
+            for func_index in indices:
+                arg = loopy.GlobalArg("coeff"+str(get), shape=coeff_shape_list[func_index], dtype="double")
+                args.append(arg)
+                get += 1
+        except ValueError:
+            pass
+
+    # Add all other cofficients
+    for v, k in builder.extra_coefficients:
+        args.append(loopy.GlobalArg(k,
+                                    shape=builder.args_extents[k],
+                                    dtype="double"))
+
+    # Arg for is exterior (==0)/interior (==1) facet or not
+    if builder.needs_cell_facets:
+        args.append(loopy.GlobalArg(builder.cell_facets_arg,
+                                    shape=(builder.num_facets, 2),
+                                    dtype=np.int8))
+
+    if builder.needs_mesh_layers:
+        args.append(loopy.TemporaryVariable("layer", shape=(), dtype=np.int32, address_space=loopy.AddressSpace.GLOBAL))
+
+    ############
+    # TODO: preprocessing of gem for removing unneccesary component tensors
+    # print("not peprocessed",traversed_gem_expr_dag)
+    # traversed_gem_expr_dag = impero_utils.preprocess_traversedgem(traversed_gem_expr_dag[0])
+    # print("preprocessed",traversed_gem_expr_dag)
+    # print(traversed_gem_expr_dag[1])
+    # traversed_gem_expr_dag=traversed_gem_expr_dag[1]
```

Comment:
So `impero_utils.preprocess_gem` would (hopefully) call `remove_componenttensors`, which would eliminate most `ComponentTensor`s, but of course it cannot eliminate `ComponentTensor`s in `Inverse` and `Factorize` nodes, so code generation for `ComponentTensor`s is still needed.